### ISSUE TITLE
MAINLOCK insufficient for protecting non-assisted cmpxchg1/4/8/16 functions

### ISSUE THREAD

**Issue by Peter-J-Jansen** [SDL-Hercules-390/hyperion#247]
`OBTAIN_MAINLOCK` / `RELEASE_MAINLOCK` does not protect against main memory changes by regular instructions such as MVI or ST and all others.

In the case of non-assisted cmpxchg1/4/8/16 functions (all of which are surrounded by such MAINLOCK protection pairs), such as when building for **Raspberry Pi (RPI)** ARM processors, the last time this caused a problem was prior to my fix in May 2015 for the TS instruction which uses cmpxchg1. That fixed solved the problem when IPL-ing z/VM on an RPI 2.

With the **RPI _4_** with 4GB now being available, and a 64 bit (unofficial) Ubuntu 18.04.3, z/OS can now also be successfully IPL'd on it. However, after some time, z/OS freezes. (It can take up to a few hours for it to occur, but is (was) quite reproducible). I suspect the problem to be with MAINLOCK not sufficiently protecting non-assisted cmpxchg# instructions.

As C11 atomics are now readily available, I coded assisted cmpxchg1/4/8 functions, and the problem of z/OS freezing on the RPI 4 appears to have gone away. (The test has been running OK for about 24 hours right now.)

&nbsp;
_**QUESTION:** Is my assumption, that `MAINLOCK` can only protect against other MAINLOCK protected main memory accesses, correct or not?_
&nbsp;

If the answer to this question is positive, then I question the usefulness of MAINLOCK completely!

Currently, MAINLOCK protects about 9 "simple" compare-and-swap style instructions; TS is just one of them. With "simple" I mean that these all end up using a single cmpxchg# function, which in the case of hardware- (C11-) assist protects itself, _also against all regular main memory access_.

Additionally, 9 other instructions are currently MAINLOCK protected : 4 MVS assist and 1 LKPG instructions (these 5 are perhaps not used anymore), 2 "complex" compare-and-swap style instructions, CSST and PLO, and 2 instructions needing 16 byte atomic store and load (STPQ and LPQ).

I think we'd need to find a better solution so that we don't need MAINLOCK anymore. Provided hardware- (C11-) assists are available, I believe this can be safely done.

_**Feedback/Discussion is strongly encouraged!**_

Cheers,

Peter

**Comment by Fish-Git** [comment]
> **QUESTION:** Is my assumption, that `MAINLOCK` can only protect against other MAINLOCK protected main memory accesses, correct or not?

Duh!&nbsp;`:)`

A lock, _any_ lock, can only protect the specific resource it was meant (designed) to protect, and only when the lock is used before accessing that resource. This is standard lock usage. As asked, your question is quite rhetorical: the answer is of course YES (which I'm sure you already knew).

> I think we'd need to find a better solution so that we don't need MAINLOCK anymore. _**Provided hardware- (C11-) assists are available,**_ I believe this can be safely done.

_THAT_ is the key: _**Provided hardware- (C11-) assists are available!**_

But what if hardware/C11 assists are _not_ available? What then? If that is the case then it sounds like we would still need MAINLOCK. Yes?

Based on what you've written (and a _brief_ (cursory) examination of the mentioned code), it sounds to me like we _cannot_ safely remove MAINLOCK because hardware/C11 assists might not be available on all supported systems. Yes?

Now what we _might_ be able to _perhaps_ do is maybe devise a new `COND_OBTAIN/RELEASE_MAINLOCK` macro that does nothing whenever hardware/C11 assists are available, and use that instead in those few specific places where `OBTAIN/RELEASE_MAINLOCK` is currently used to wrap a `cmpxchg` call, such as the `compare_and_swap` instruction for example:

```diff
--- hyperion-1/general1.c	2019-08-07 03:23:15.530798900 -0700
+++ hyperion-0/general1.c	2019-09-16 03:48:49.722035400 -0700
@@ -1898,13 +1898,13 @@
     old = CSWAP32(regs->GR_L(r1));
 
     /* Obtain main-storage access lock */
-    OBTAIN_MAINLOCK(regs);
+    COND_OBTAIN_MAINLOCK(regs);
 
     /* Attempt to exchange the values */
     regs->psw.cc = cmpxchg4 (&old, CSWAP32(regs->GR_L(r3)), main2);
 
     /* Release main-storage access lock */
-    RELEASE_MAINLOCK(regs);
+    COND_RELEASE_MAINLOCK(regs);
 
     /* Perform serialization after completing operation */
     PERFORM_SERIALIZATION (regs);
```

This would eliminate the overhead of obtaining/releasing a lock (MAINLOCK) for those specific cases where doing so wasn't actually necessary (because the `cmpxchg4` function itself, by virtue of it using hardware/C11 assists, would therefore eliminate the need for using such a lock), but _**only**_ in such cases would it be safe. In other situations it might not be. In other cases we might still need to use MAINLOCK. I don't know. I haven't studied the code or thought about things enough yet.

But that's my initial feedback off the top of my head.

Anyone else want to jump in with their thoughts?

**Comment by Fish-Git** [comment]
I'm also curious what your proposal is for resolving the actual problem as described by the title of this GitHub Issue: _"MAINLOCK insufficient for protecting non-assisted cmpxchg1/4/8/16 functions"_.

If MAINLOCK isn't good enough for protecting non-assisted cmpxchg1/4/8/16 functions, then what do you propose we do to fix that? How do you propose we fix things so that non-assisted cmpxchg1/4/8/16 functions _are_ protected? That is to say, how do we reliably protect cmpxchg1/4/8/16 function calls when hardware/C11 assists are _not_ available?

Right now we're using MAINLOCK (which you're saying isn't good enough). But if using MAINLOCK in such situations isn't good enough, then what is? What's the fix?

**Comment by Peter-J-Jansen** [comment]
Dear Fish,

Thanks for your feedback. 

In the mean time, I've been quite busy researching this some more, and the first thing I now have to do is to correct myself somewhat and be more precise. What I referred to as "compare-and-swap" style instructions is probably more correctly referred to by the Principles of Operations (PoO) wording, which calls this instructons with "Interlocked-Update References" (IUR). The SA22-7832-11 PoO lists them on pg. 5-124, and it shows that they _only protect against themselves_. I.e. they do not need to protect against all other instructions, such as the MVI and the ST instructions I mentioned initially. But (some of ?) the Interlocked Access Facility (IAF) 1 and 2 instructions are mentioned, and _none of these are currently MAINLOCK protected_! The IAF instructions rely on the C11 atomics only (macro H_ATOMIC_OP).

The reason we've been getting away with this is that on our main processor architecture, amd64 (i.e. the current Intel's), _all_ of these IUR instructions are assisted, for Windows (thanks to you), as well as for Linux. OK, with the exception of 2 instructions : CDSG (in esame.c) and CSST (in general1.c), which rely on cmpxchg16, for which currently no assist is available.       

In order to add support for my newest platform, a 4GB RPI 4, I coded cmpxchg1/4/8 assists using GCC's C11 atomics, which appear to work fine, also for MacOS. As a result, z/OS 2.3 (as well as z/VM 6.3) run quite stable, also on MacOS and the RPI 4.  Unfortunately, cmpxchg16 poses problems.  The GCC C11 atomics generate some library calls for cmpxchg16, for which I have to add the GCC option "-latomic", but the resulting build does not work yet.  For the amd64 platform I was able to code a "cmpxchg16b" based assist (thanks to your FIXME hint in machdep.h). For the RPI 4 platform I discovered that it's CPU is ARMv8, which does not support the "casp" instruction (only ARMv8.1 does). So this cmpxchg16 will need a different approach, I don't know yet how to do that.

One instruction which is remarkably missing from the PoO list on pg. 5.124 is the Perform Locked Operation (PLO) instruction. The PLO instruction (in general.2) is MAINLOCK protected, and this is just fine, as it only has to protect against other PLO's on other CPU's.  This MAINLOCK can stay there unchanged ! However, I'd rename that MAINLOCK to MAINLOCK_PLO or MAINLOCK_UNCONDITIONAL.

The STPQ and LPQ instructions are _not_ listed on the PoO list on pg. 5.124 but nevertheless these too should, in my opinion, be implemented using newly to be created atomic store16 and load16 functions.  MAINLOCK here does not help at all. 

If all of the above assists would be available, then the current MAINLOCK macro could be modified so that if assists are available for all of the cmpxchg1/4/8 _and_ cmpxchg16, _and_ the MVS Assists are not in the build, _and_ the LKPG instruction is not in the build, that under these conditions, the MAINLOCK macro could do effectively nothing.  This MAINLOCK could thus become a conditional one.  All existing use of MAINLOCK could remain as is, only it's definitions (OBTAIN_... and RELEASE_... in hmacros.h) would need adaptation.  The end result would avoid allbut the PLO MAINLOCK, so we'd gain a bit of performance.  

Theorethically, the IAF instructions in the PoO list on pg. 5.124 should also need to be MAINLOCK protected, and I think it is (one or more of) these that caused my RPI platform problems with the TS and some other compare-and-swap instructions, prior to my assists for the ARM processor of the RPI.

What do you think Fish ?  And yes, I'm of course also very much interested what our other SDL Hercules team players think about this all !

Cheers,

Peter

**Comment by Peter-J-Jansen** [comment]
Fish, as a side note : how do you get these nice colors in your github replies ?

Cheers,

Peter

**Comment by Fish-Git** [comment]
> Fish, as a side note: how do you get these nice colors in your github replies?

You mean the diff? (patch?) Simple! Just specify "diff" (or "patch"?) as the language code in your fenced code block:

\`\`\`diff
(patch statements would go here)
(patch statements would go here)
(patch statements would go here)
\`\`\`

\`\`\`C
(source code snippet would go here...)
(source code snippet would go here...)
(source code snippet would go here...)
\`\`\`

etc...

GitHub supports many different languages. I'm not sure if there's any web page anywhere that lists _all_ of the languages that it supports, but I only ever use 'C' and patch/diff.

**NOTE:** the above _will_ be properly colored when you actually do it for real, but the above syntax examples aren't colored because I escaped the triple back ticks so you could see the actual syntax.

Ref:

* [Mastering Markdown](https://guides.github.com/features/mastering-markdown/)
* [Markdown Syntax / GitHub Flavored Markdown CHEAT SHEET](https://guides.github.com/pdfs/markdown-cheatsheet-online.pdf)

&nbsp;
p.s. You are aware that any of us (SDL Team) can "edit" anyone else's posts, right? I edit people's posts all the time so that they're better formatted and thus easier to read.

**Comment by Peter-J-Jansen** [comment]
Thanks for the diff and C tips Fish !

I have a first version ready of an ARMv8 intrinsic for cmpchgx16, so I can start testing on my RPI 4 platform for an SDL-Hercules with effectively only the PLO instruction remaining MAINLOCK-ed. I'll keep you informed. 

I'm also thinking about rigging up Hercules tests for these "compare-and-swap" instructions, but ... that is not goind to be easy, as the test program would have to be multitasking between 2 CPU's.  That may prove to above my bare metal HLASM abilities ...

Cheers,

Peter

**Comment by Fish-Git** [comment]
> What I referred to as "compare-and-swap" style instructions should probably be more correctly referred to [as] **"Interlocked-Update References" (IUR)** type instructions.

Okay.

> they _only protect against themselves_. (I.e. they do _not_ need to protect against all other instructions such as the `MVI` and the `ST` instructions that I initially mentioned.

Okay.

> But [...] the **Interlocked Access Facility (IAF) 1 and 2** instructions are [also] mentioned, and _**none of these are currently MAINLOCK protected in Hercules!**_ (The IAF instructions in Hercules currently rely _solely_ on C11 atomics, i.e. macro `H_ATOMIC_OP`.)

Well _that's_ a problem that we definitely need to fix!

> The reason we've been getting away with this is [...] all of these IUR instructions are assisted for both Windows [...] as well as for Linux [...]

Okay.

> ... _except for_ ... 2 instructions: `CDSG` (in esame.c) and `CSST` (in general1.c), both of which rely on `cmpxchg16` which is currently _unassisted_.)

So we need to definitely code an _assisted_ `cmpxchg16` macro in `machdep.h` that works across the board (i.e. for both Linux and Windows and other platforms too). To do that we're probably going to need to add some code that checks for the availability of the x86_64 `cmpxchg16b` host instruction so we can use that in our assist. **`(*)`**


> The GCC C11 atomics generate some library calls for `cmpxchg16` ...

It's probably calling some function or performing some test to check to see if the `cmpxchg16b` instruction is available or not first before attempting to use it (for the likely reason as mentioned in my footnote further below).

> (I would however recommend renaming it to maybe `MAINLOCK_PLO` or `MAINLOCK_UNCONDITIONAL` instead.)

Agreed.

> Theoretically however, the **IAF** instructions in the PoO list on pg. 5.124 would _also_ need to be MAINLOCK protected as well,

Absolutely! (based on what you've said thus far) As I said, _that **definitely**_ needs to be fixed!

> _What do you think Fish?_

I think you've got a good handle on things, Peter, and have complete trust in your ability to resolve this issue. If you need any help with anything please don't hesitate to ask.

> _**(And yes, I'm of course also very much interested what our other SDL Hercules team players think about all of this!)**_

ME TOO!!

Guys?&nbsp;`:)`

___
**`(*)`** I'm not sure how such works on other platforms, but on Windows, the documentation clearly states: _"If you run code that uses this intrinsic on hardware that does not support the cmpxchg16b instruction, the results are unpredictable."_, which I interpret as meaning your program will crash. That is to say, using the `_InterlockedCompareExchange128` compiler intrinsic simply reduces to a `cmpxchg16b` x86_64 host instruction, so if the processor that Hercules is running on is one that doesn't support the `cmpxchg16b` instruction, then it will crash. Thus we're going to need to develop some type of "configure.ac" test that checks to see if the instruction is available and set a #define constant that `machdep.h` can then check. (Not sure how we're going to do that on Windows since Windows doesn't do any type of pre-build "configure.ac" processing, but I'm sure we can figure something out.) And we're of course going to need to code a reliable method to accomplish the same thing for the case where it's not available. (That is to say, if the `cmpxchg16b` instruction is _not_ available, our new `cmpxchg16` macro in `machdep.h` is going to need to somehow reliably accomplish the same thing using other instructions.)

**Comment by Fish-Git** [comment]
_THANK YOU Peter! &nbsp; Great work!_

> ... I simply could _not_ manage to get the MSVC `_InterlockedCompareExchange128` intrinsic to work.

> I am prepared to think that perhaps there is an MSVC bug involved here, but I dare not say that as I am too unfamiliar with MSVC intrinsics.

> Perhaps Fish can rescue me here?

I'll certainly try!

There is a chance that the `_InterlockedCompareExchange128` intrinsic has simply never been used by anyone before (or if so, not much), so your suspicion that there's a bug in MSVC is entirely reasonable. But let me give it a try myself first, and I'll let you know whether I can manage to get it to work or not.

To maybe expedite things a little bit, can you send me your `CDSG` test? That might help speed things up for me. Thanks.

> And I changed to `OBTAIN_ / RELEASE_MAINLOCK` around the `PLO` instruction to `OBTAIN_ / RELEASE_MAINLOCK_UNCONDITIONAL`, so that _all other MAINLOCK usage_ can become conditional when all `cmpxchg1 / 4 / 8 / 16` are assisted.

I'm not seeing that in your patch. I just applied it, and the `PLO` instruction remains unpatched. Simple oversight perhaps?

___
p.s. IMO your patch (diff) is little too large to be pasting directly into a GitHub comment. You and others may of course feel differently, but IMO when I have a patch I want to share that's over, say, 60-75 lines or so (approximately), I zip it into a .zip file and then simply attach that .zip file to the comment instead. (Or rename it to .txt and attach that text file).

But as I said, that's just my own feelings on the matter. Do whatever you want! I'm not your _(or anyone else's!)_ boss.

&nbsp; &nbsp; &nbsp; &nbsp; `;-)`

**Comment by Fish-Git** [comment]
> I'm not seeing that in your patch. I just applied it, and the `PLO` instruction remains unpatched. Simple oversight perhaps?

_**NEVER MIND!!**_

I just noticed the `PLO` instruction is actually in `general2.c`, not `plo.c`!

The code in `plo.c` is just the helper routines that the `PLO` instruction function in `general2.c` calls into. &nbsp;_(Doh!)_

Sorry about that. &nbsp; &nbsp; `%-þ`

**Comment by Peter-J-Jansen** [comment]
The CDSG test files (the usual .core, .asm, .list and .tst) I did dare committing as they are a pure addition and also work without any source changes, also against a non-assisted cmpxchg16.

Thanks Fish, also for your feedback !

Cheers,

Peter

**Comment by Peter-J-Jansen** [comment]
This can now be closed as of commit e0052fd. Many thanks to Fish for all the help he provided.

Cheers,

Peter
